src/main.rs

```diff
@@ -3,28 +3,19 @@
 
 mod algo;
 
-use core::mem;
 use core::mem::MaybeUninit;
 
 use self::algo::*;
 
 fn find_func<T>(tag: [u8; 2]) -> T {
-    let tag = u16::from_le_bytes(tag);
-
+    let tag = u16::from_le_bytes(tag) as u32;
+    type RomTableLookupFn = unsafe extern "C" fn(table: *const u16, code: u32) -> usize;
     unsafe {
-        let mut entry = *(0x00000014 as *const u16) as *const u16;
-        loop {
-            let entry_tag = entry.read();
-            if entry_tag == 0 {
-                panic!("Func not found");
-            }
-            entry = entry.add(1);
-            let entry_addr = entry.read();
-            entry = entry.add(1);
-            if entry_tag == tag {
-                return mem::transmute_copy(&(entry_addr as u32));
-            }
-        }
+        let lookup_func = core::ptr::read(0x0000_0018 as *const u16) as usize;
+        let lookup_func: RomTableLookupFn = core::mem::transmute(lookup_func);
+        let table = core::ptr::read(0x0000_00014 as *const u16) as usize;
+        let result = lookup_func(table as *const u16, tag);
+        core::mem::transmute_copy(&result)
```

Comment:
This will return 0 instead of panicking if an unknown tag is used. Just mentioning. I think it's fine, this function is very unsafe if used incorrectly anyway. If that was meant to be included in a larger code base, I'd ask for safety comments. But I think here it's ok. The whole program is using rust not as a safe language, but as a more convenient alternative to assembly.

Reply:
That's a good point, and as the ROM contents are defined outside the program I should probably trap that.

Reply:
It would be cleaner, but the practical impact would be very small: If you use a flash algo on the wrong chip, chances for a satisfying result are very small either way.

Reply:
The init function can return an error, so it makes sense to do that instead.